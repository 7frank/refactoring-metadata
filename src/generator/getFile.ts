import { Project } from "ts-morph";

export function getFile(src, data: string) {
  const project = new Project({ useInMemoryFileSystem: true });
  //  const fs = project.getFileSystem();

  const sourceFile = project.createSourceFile(src, data);
  project.addDirectoryAtPathIfExists("../fixtures");

  sourceFile.saveSync();
  return sourceFile;
}
