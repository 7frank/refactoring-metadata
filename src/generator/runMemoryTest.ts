import * as ts from "typescript";

/**
 * we probably ned to run a whole compilation of complex sources in a memory or ramdisk file system
 * @param source
 */
export async function runMemoryTest(source: string) {
  let result = ts.transpileModule(source, {
    compilerOptions: { module: ts.ModuleKind.CommonJS },
  });

  try {
    const testResult = eval(result.outputText);
    return JSON.stringify(testResult, null, "  ");
  } catch (e) {
    const E = new Error(
      "Error on evaluating:" +
        e.message +
        "\n\n Javascript:" +
        result.outputText +
        "\n\n"
    );
    E.stack = e.stack;
    throw E;
  }
}
