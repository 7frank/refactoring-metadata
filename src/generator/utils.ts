import * as ts from "typescript";
import * as path from "path";
import type { InterfaceDefinition } from "../xray/types";
import { Project } from "ts-morph";

var jsonic = require("jsonic");

var traverse = require("traverse");

/**
 * Note run ts-node src/index2
 * TODO create separate package with unit tests for cases
 * initial setup: parse file
 *
 *  1 write tests that extract jsx function
 *  2 run it through a template that generates a jest test
 *  3 have a function that executes the test directly
 *  4 ( use 3 together with an option to only generate valid test from file)
 *
 *  1 write test that extracts potential unit test candidates
 *  2 unit test template
 *  3 createtest data & compilefile containing function % execute it
 *  4 create unit test from 2 & 3
 *
 *
 * x we need recursion to resolve "function" params
 */

export function loadFile(filepath: string, project?: Project) {
  project =
    project ??
    new Project({
      compilerOptions: {
        jsx: 2,
      },
    });
  return project.addSourceFileAtPath(filepath);
}

/**
 * this is a really rough draft of things we might wanna do
 */
export function createSimpleTests(
  element: InterfaceDefinition,
  name: string,
  src: string
) {
  const { __return__, ...props } = element as any;

  /**
   * TODO don't traverse initially as our type might be an array of types via "|" where we'd have to pick one
   */

  let dummyProps = createDummyProps(props);

  // todo run prettier before out
  if (__return__ == "JSX.Element") {
    return createReactTest({ name, path: "", props: dummyProps });
  } else if (__return__ != "JSX.Element") {
    // TODO random values obviously will fail BUT we actually can compile the file
    //  and its dependencies and run it with random input and
    const returnValue = createDummyProps(jsonic(__return__ as string));

    const test = createRandomFunctionCall({
      name,
      path: extractDummyPath(src),
      params: createDummyParams(props),
    });
    runMemoryTest(test);

    return createUnitTest({
      name,
      path: "",
      props: dummyProps,
      returnValue,
    });
  }
}

function extractDummyPath(src) {
  const f = path.basename(src);

  return f.split(".").slice(0, -1).join(".");
}

function createDummyProps(props) {
  const dummyProps = Object.entries(props).map(([name, prop]) => {
    traverse(prop).forEach(function (x) {
      if (!this.isLeaf) return;

      const key = this.node;

      if (key == "number") this.update(Math.random());
      if (key == "string") this.update("randomString");
      if (this.node == "boolean") this.update(true);

      if (key == "number[]") this.update([Math.random(), Math.random()]);
      if (key == "string[]") this.update(["'randomString'", "'randomStrin2'"]);
      if (key == "boolean[]") this.update([true, false]);
    });

    const stringifiedProp = JSON.stringify(prop, null, "  ");
    return `${name} = {${stringifiedProp}}`;
  });
  return dummyProps;
}

function createDummyParams(props) {
  traverse(props).forEach(function (x) {
    if (!this.isLeaf) return;

    const key = this.node;

    if (key == "number") this.update(Math.random());
    if (key == "string") this.update("randomString");
    if (this.node == "boolean") this.update(true);

    if (key == "number[]") this.update([Math.random(), Math.random()]);
    if (key == "string[]") this.update(["'randomString'", "'randomStrin2'"]);
    if (key == "boolean[]") this.update([true, false]);
  });
  return Object.values(props).map((entry) => JSON.stringify(entry, null, "  "));
}

function createReactTest({
  name,
  path,
  props,
}: {
  name: string;
  path: string;
  props: string[];
}) {
  const cName = name;

  const out = `

    /**
     * test for ${cName}
     */
  import {${cName}}  from './Button.tsx'
    

    it('renders correctly', () => {
       // arrange
       const randomTestProps=${props.join(" ")}
   
       // act
        const tree = renderer
          .create(<${cName} {...randomTestProps} ></${cName}>)
          .toJSON();
        
        // assert
        expect(tree).toMatchSnapshot();
      });
    `;

  return out;
}

function createUnitTest({
  name,
  path,
  props,
  returnValue,
}: {
  name: string;
  path: string;
  props: string[];
  returnValue: string[];
}) {
  const cName = name;

  const out = `

    /**
     * test for ${cName}
     */
  import {${cName}}  from './Button.tsx'
    

    it('renders correctly', () => {
       // arrange
       const randomTestProps=${props.join(" ")}
   
       // act
      const result= ${cName}(randomTestProps)
        
        // assert
        expect(result).toBeEqual(${returnValue.join(" ")});
      });
    `;

  return out;
}

function createRandomFunctionCall({
  name,
  path,
  params,
}: {
  name: string;
  path: string;
  params: string[];
}) {
  const cName = name;

  return `

   import {${cName}}  from './${path}'
 
   const randomTestProps=[${params.join(",")}];
  
   ${cName}(...randomTestProps)
      
`;
}

/**
 * we probably ned to run a whole compilation of complex sources in a memory or ramdisk file system
 * @param source
 */
async function runMemoryTest(source: string) {
  //  const source = 'let x: string  = \'string\'';

  let result = ts.transpileModule(source, {
    compilerOptions: { module: ts.ModuleKind.CommonJS },
  });

  console.log("/------------------------------------/");
  console.log(source);
  console.log("/------------------------------------/");
  console.log(JSON.stringify(result));
  console.log("/------------------------------------/");
  console.log(result.outputText);
  console.log("/------------------------------------/");
  try {
    const testResult = eval(result.outputText);
    console.log("testResult:");
    console.log(JSON.stringify(testResult));
  } catch (e) {
    console.log("Error on Eval", e);
  }

  console.log("/------------------------------------/");
}
