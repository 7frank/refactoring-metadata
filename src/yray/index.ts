import {
    Project,
    InterfaceDeclaration,
    SourceFile,
    ImportDeclaration,
    FunctionDeclaration,
  } from "ts-morph";
  import { ParameterDeclaration } from "typescript";
  // eslint-disable-next-line
  import type {
    TypenameToUnresolvedRefsMap,
    InterfaceDefinitions,
    InterfaceDefinition,
    Filepath,
  } from "tsx-ray/dist/types";

  import {
    PrimitiveType,
    ArrayType,
    ParsedType,
    UnionType,
    ObjectType,
  } from "tsx-ray/dist/types";

  import {
    merge,
    isImport,
    getNameFromType,
    convertToArrayRepresentation,
    convertToPrimitiveRepresentation,
    removeQuotesIfLiteral,
  } from "tsx-ray/dist/utils";
  
  export const compileFile = (filepath: Filepath, outDir: string) => {
    const project = new Project({
      compilerOptions: {
        outDir,
        declaration: false,
        jsx: 2,
      },
    });
    const sourceFile = project.addSourceFileAtPath(filepath);
    sourceFile.emitSync();
  };
  
  export const compileSourceSync = (sourceFile: SourceFile) =>
    sourceFile.emitSync();
  
  export const extractInterfaces = (sourceFile: SourceFile) =>
    parseInterfacesFromSourceFile(sourceFile);
  
  export const getDefaultImports = (sourceFile: SourceFile) => {
    return sourceFile
      .getImportDeclarations()
      .filter((imp) => imp.getDefaultImport());
  };
  
  export const extractFunctionsFromFile = (sourceFile:SourceFile) => {
    // const project = new Project({
    //   compilerOptions: {
    //     jsx: 2,
    //   },
    // });
    // const sourceFile = project.addSourceFileAtPath(filepath);
  
    const otherInterfaces = parseInterfacesFromSourceFile(sourceFile);
  
    return parseFunctionsFromSourceFile(sourceFile, otherInterfaces);
    //  return parseInterfacesFromSourceFile(sourceFile);
  };
  
  /**
   * exclude specific imports from parsing any further
   */
  function importFilter(src: SourceFile, array: string[] = []) {
    const path = src.getFilePath();
  
    const found = array.some(function (entry) {
      const reg = new RegExp(`[\\\\/]${entry}[\\\\/]`);
      return reg.test(path);
    });
  
    return !found;
  }
  
  const parseInterfacesFromSourceFile = (
    sourceFile: SourceFile
  ): InterfaceDefinitions => {
    const interfaces = sourceFile.getInterfaces();
    const imports = sourceFile.getImportDeclarations();
  
    const getParsedInterfacesFromImports = (imports: ImportDeclaration[]) => {
      if (imports.length > 0) {
        const importedSources = imports
          .map((i) => i.getModuleSpecifierSourceFile())
          .filter(Boolean) as SourceFile[];
        const importedParsedInterfaces = importedSources
          .filter((src) => importFilter(src))
          .map(parseInterfacesFromSourceFile);
        return importedParsedInterfaces;
      }
      return [];
    };
  
    const interfaceDefinitions: InterfaceDefinitions = {};
    const unresolvedTypes: Set<string> = new Set();
    const typeReferences: TypenameToUnresolvedRefsMap = {};
  
    const importedInterfaces = getParsedInterfacesFromImports(imports);
  
    const allInterfaceDefinitions =
      importedInterfaces.length > 0
        ? importedInterfaces.reduce(
            (acc, current) => merge(acc, current),
            interfaceDefinitions
          )
        : interfaceDefinitions;
  
    const parseInterfaceProperties = (intf: InterfaceDeclaration) => {
      const properties: InterfaceDefinition = {};
      for (const property of intf.getProperties()) {
        const type = property.getType();
  
        if (type.isArray()) {
          const typeArgs = type.getTypeArguments();
          const arrayElementType = convertToArrayRepresentation(typeArgs[0]);
          properties[property.getName()] = arrayElementType;
        } else if (type.isBoolean()) {
          properties[property.getName()] = convertToPrimitiveRepresentation(type);
        } else if (type.isUnion()) {
          const unionTypes = type
            .getUnionTypes()
            .map(removeQuotesIfLiteral) as PrimitiveType[];
  
          properties[property.getName()] = unionTypes;
        } else if (type.isInterface()) {
          const rawText = type.getText();
          const nameOfType = isImport(rawText) ? getNameFromType(type) : rawText;
  
          unresolvedTypes.add(nameOfType);
  
          const unresolvedPropertyRef = {
            ref: properties,
            name: property.getName(),
          };
  
          if (!typeReferences[nameOfType]) {
            typeReferences[nameOfType] = [];
          }
  
          typeReferences[nameOfType].push(unresolvedPropertyRef);
        } else {
          properties[property.getName()] = convertToPrimitiveRepresentation(type);
        }
      }
      return properties;
    };
  
    for (const intf of interfaces) {
      allInterfaceDefinitions[intf.getName()] = parseInterfaceProperties(intf);
    }
  
    for (const unresolvedType of Array.from(unresolvedTypes)) {
      if (!allInterfaceDefinitions[unresolvedType]) {
        console.warn(`No definition found for ${unresolvedType}`);
      }
  
      const unresolvedReferences = typeReferences[unresolvedType];
  
      for (const unresolvedPropertyRef of unresolvedReferences) {
        const ref = unresolvedPropertyRef.ref;
        ref[unresolvedPropertyRef.name] = allInterfaceDefinitions[unresolvedType];
      }
    }
  
    return allInterfaceDefinitions;
  };
  
  // let importSelf = true;
  
  const parseFunctionsFromSourceFile = (
    sourceFile: SourceFile,
    otherInterfaces: InterfaceDefinitions
  ): InterfaceDefinitions => {
    //const interfaces = sourceFile.getInterfaces();
    const imports = sourceFile.getImportDeclarations();
  
    const functions = sourceFile.getFunctions();
  
    // const rFunction=sourceFile.getFunctions()[1]
    // const testReturnType=rFunction.getReturnType()
    // // will be "JSX.Element" our react function
    // const result2= convertToPrimitiveRepresentation(testReturnType)
    // const properties: InterfaceDefinition = {};
    // const testParameterTypes=rFunction.getParameters().forEach(p => properties[p.getName()]=convertToPrimitiveRepresentation(p.getType()))
  
    const getParsedInterfacesFromImports = (imports: ImportDeclaration[]) => {
      if (imports.length > 0) {
        const importedSources = imports
          .map((i) => i.getModuleSpecifierSourceFile())
          .filter(Boolean) as SourceFile[];
  
        // add this sourcefile to import local stuff
        // if (importSelf) {
        //   importSelf = false;
        //   importedSources.push(sourceFile);
        // }
        const importedParsedInterfaces = importedSources
          .filter((src) => importFilter(src, ["prop-types", "csstype"]))
          .map((sourceFile) =>
            parseFunctionsFromSourceFile(sourceFile, otherInterfaces)
          );
        return importedParsedInterfaces;
      }
      return [];
    };
  
    const interfaceDefinitions: InterfaceDefinitions = {};
    const unresolvedTypes: Set<string> = new Set();
    const typeReferences: TypenameToUnresolvedRefsMap = {};
  
    const importedInterfaces = getParsedInterfacesFromImports(imports);
  
    const allInterfaceDefinitions =
      importedInterfaces.length > 0
        ? importedInterfaces.reduce(
            (acc, current) => merge(acc, current),
            interfaceDefinitions
          )
        : interfaceDefinitions;
  
    const parseFunctionInterfaceProperties = (intf: FunctionDeclaration) => {
      const properties: InterfaceDefinition = {};
  
      const params = intf.getParameters();
  
      // add partial return type interface
      const ret = {
        getName: () => "__return__",
        getType: () => intf.getReturnType(),
      };
      params.push(ret as any);
  
      for (const property of params) {
        const type = property.getType();
  
        if (type.isArray()) {
          const typeArgs = type.getTypeArguments();
          const arrayElementType = convertToArrayRepresentation(typeArgs[0]);
          properties[property.getName()] = arrayElementType;
        } else if (type.isBoolean()) {
          properties[property.getName()] = convertToPrimitiveRepresentation(type);
        } else if (type.isUnion()) {
          const unionTypes = type
            .getUnionTypes()
            .map(removeQuotesIfLiteral) as PrimitiveType[];
  
          properties[property.getName()] = unionTypes;
        } else if (type.isInterface()) {
          const rawText = type.getText();
          const nameOfType = isImport(rawText) ? getNameFromType(type) : rawText;
  
          unresolvedTypes.add(nameOfType);
  
          const unresolvedPropertyRef = {
            ref: properties,
            name: property.getName(),
          };
  
          if (!typeReferences[nameOfType]) {
            typeReferences[nameOfType] = [];
          }
  
          typeReferences[nameOfType].push(unresolvedPropertyRef);
        } else {
          properties[property.getName()] = convertToPrimitiveRepresentation(type);
        }
      }
      return properties;
    };
  
    for (const intf of functions) {
      allInterfaceDefinitions[intf.getName()] = parseFunctionInterfaceProperties(
        intf
      );
    }
  
    /**
     * FIXME props and look for pattern like import(\"/home/frank/playground-my-plugin/src/ImportedProp\").ImportedProp
     * and replace those with otherInterfaces[ImportedProp]
     */
  
    for (const unresolvedType of Array.from(unresolvedTypes)) {
      let resolved = allInterfaceDefinitions[unresolvedType];
  
      if (!resolved) {
        console.warn(
          `No definition found for ${unresolvedType} attempting backup`
        );
  
        if (unresolvedType == "JSX.Element") {
          // @ts-ignore
          resolved = "JSX.Element";
        } else if (!otherInterfaces[unresolvedType]) {
          console.warn(`No definition found in backup for ${unresolvedType}`);
        } else {
          resolved = otherInterfaces[unresolvedType];
        }
      }
  
      const unresolvedReferences = typeReferences[unresolvedType];
  
      for (const unresolvedPropertyRef of unresolvedReferences) {
        const ref = unresolvedPropertyRef.ref;
        ref[unresolvedPropertyRef.name] = resolved;
      }
    }
  
    return allInterfaceDefinitions;
  };
  
  export { PrimitiveType, ArrayType, ObjectType, ParsedType, UnionType };