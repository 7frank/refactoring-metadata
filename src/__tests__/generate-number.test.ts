import { extractFunctionsFromFile } from "../yray";
import { createSimpleTests, loadFile } from "../generator/utils";
import { getFile } from "../generator/getFile";

require("jest-specific-snapshot");

describe("inc test generator", () => {
  const src = ""; //no path available

  const sourceFile = getFile(
    "file.ts",
    `
  export function inc(val: number) {
    return val++;
  }
`
  );

  const result = extractFunctionsFromFile(sourceFile);

  it("generates interface definitions", async () => {
    expect(result).toMatchSnapshot("json interface payload");
  });

  it("creates test naive cases", async () => {
    for await (const [name, element] of Object.entries(result)) {
      const resultingUnitTestCode = await createSimpleTests(
        element,
        name,
        src
      ).catch((e) => {
        expect(
          e.message + "\n\n" + sourceFile.getFullText()
        ).toMatchSpecificSnapshot(
          "./__snapshots__/mem-inc.ts/" + name + ".snap"
        );
      });
      expect(resultingUnitTestCode).toMatchSpecificSnapshot(
        "./__snapshots__/mem-inc.ts/" + name + ".snap"
      );
    }
  });
});
