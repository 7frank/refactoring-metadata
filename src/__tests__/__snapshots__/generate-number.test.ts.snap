// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inc test generator generates interface definitions: json interface payload 1`] = `
Object {
  "inc": Object {
    "__return__": "number",
    "val": "number",
  },
}
`;
